In [7]:
# from google.colab import drive
# drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
!nvidia-smi

Mon Apr 24 09:11:04 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   49C    P8    10W /  70W |      0MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [4]:
!git clone https://github.com/elmizb/alignn_lightning.git
%cd ./alignn_lightning

Cloning into 'alignn_lightning'...
remote: Enumerating objects: 66, done.
remote: Counting objects: 100% (66/66), done.
remote: Compressing objects: 100% (47/47), done.
remote: Total 66 (delta 34), reused 46 (delta 19), pack-reused 0
Unpacking objects: 100% (66/66), 123.76 KiB | 2.43 MiB/s, done.
/content/alignn_lightning


import pytorch and check the CUDA version  
If you want to use pytorch geometric, uncomment the below 5 lines to install.

In [3]:
import torch

def format_pytorch_version(version):
  return version.split('+')[0]

TORCH_version = torch.__version__
TORCH = format_pytorch_version(TORCH_version)

def format_cuda_version(version):
  return 'cu' + version.replace('.', '')

CUDA_version = torch.version.cuda
CUDA = format_cuda_version(CUDA_version)

# !pip install torch-scatter     -f https://pytorch-geometric.com/whl/torch-{TORCH}+{CUDA}.html -qqq
# !pip install torch-sparse      -f https://pytorch-geometric.com/whl/torch-{TORCH}+{CUDA}.html -qqq
# !pip install torch-cluster     -f https://pytorch-geometric.com/whl/torch-{TORCH}+{CUDA}.html -qqq
# !pip install torch-spline-conv -f https://pytorch-geometric.com/whl/torch-{TORCH}+{CUDA}.html -qqq
# !pip install torch-geometric -qqq

dgl library installation  
It is not needed if you use pytorch_geometric.

In [4]:
!pip install  dgl -f https://data.dgl.ai/wheels/{CUDA}/repo.html -qqq
!pip install  dglgo -f https://data.dgl.ai/wheels-test/repo.html -qqq

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 265.6/265.6 MB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.5/63.5 kB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 109.5/109.5 kB 5.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.2/45.2 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 29.4/29.4 MB 30.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.2/91.2 kB 10.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.4/52.4 kB 5.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.8/78.8 kB 9.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 4.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 68.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 519.4/519.4 kB 47.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 570.5/570.5 kB 52.6 MB/s eta 0:

Installation for other libraries

In [5]:
!pip install pytorch-lightning -qqq
!pip install pandarallel -qqq
!pip install pymatgen -qqq
!pip install jarvis-tools -qqq

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 718.6/718.6 kB 11.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 519.2/519.2 kB 41.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 59.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 269.3/269.3 kB 29.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.8/158.8 kB 20.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.2/114.2 kB 11.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 kB 4.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 68.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 561.4/561.4 kB 37.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.0/67.0 kB 8.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 74.2/74.2 kB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.4/98.4 kB 9.6 MB/s eta 

In [8]:
import json
import torch
import pandas as pd
from sklearn.metrics import mean_squared_error
import pytorch_lightning as pl
from pytorch_lightning.callbacks.early_stopping import EarlyStopping
# from pytorch_lightning.callbacks import StochasticWeightAveraging
# from pytorch_lightning.tuner import Tuner
from alignn_pl.data_pl import ALIGNNDataModule
from alignn_pl.config import ALIGNNConfig as config
# from alignn_pyg.model_pl import ALIGNN # ALIGNN model with pytorch_geometric
from alignn_pl.alignn import ALIGNN # model with dgl

### download the jarvis datasets (jarvis_dft3d_2021)  
In the original paper of ALIGNN, the version of jarvis_dft datasets is 2021.  
There is useful method in jarvis-tools to download the datasets but it can't do it for "jarvis_dft3d_2021"

In [21]:
!mkdir data
!wget https://figshare.com/ndownloader/files/26808914 -O ./data/jarvis_dft3d_2021.zip

--2023-04-24 02:33:06--  https://figshare.com/ndownloader/files/26808914
Resolving figshare.com (figshare.com)... 54.74.188.67, 54.170.24.133, 2a05:d018:1f4:d003:692:43c5:e3ee:104a, ...
Connecting to figshare.com (figshare.com)|54.74.188.67|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/26808914/d33122021.json.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230424/eu-west-1/s3/aws4_request&X-Amz-Date=20230424T023306Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=17bcccb44216b9f9879ed4a01a1f205c34176ca2fad4b60b550584cf0d085588 [following]
--2023-04-24 02:33:06--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/26808914/d33122021.json.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230424/eu-west-1/s3/aws4_request&X-Amz-Date=20230424T023306Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=17bcccb44216b9f9879ed4a01a1f205c

In [28]:
!unzip ./data/jarvis_dft3d_2021.zip -d ./data/

Archive:  ./data/jarvis_dft3d_2021.zip
  inflating: ./data/d3-3-12-2021.json  


In [29]:
with open("./data/d3-3-12-2021.json", 'r') as f:
    data_json=json.load(f)
ds = pd.DataFrame(data_json)

In [30]:
ds.head(2)

jid spg_number spg_symbol     formula formation_energy_peratom  \
0  JVASP-44377        186     P6_3mc  Li2Cr3SbO8                 -2.07367   
1  JVASP-14997        191     P6/mmm       YbNi5                 -0.32266   

        func optb88vdw_bandgap  \
0  OptB88vdW             0.362   
1  OptB88vdW                 0   

                                               atoms slme magmom_oszicar  ...  \
0  {'lattice_mat': [[2.9521561771454454, -5.11328...   na             na  ...   
1  {'lattice_mat': [[2.406036572951789, -4.167378...   na          1.206  ...   

  poisson                                          raw_files nat  \
0      na  [OPT-LOPTICS,JVASP-44377.zip,https://ndownload...  28   
1    0.29  [FD-ELAST,JVASP-14997.zip,https://ndownloader....   6   

  bulk_modulus_kv shear_modulus_gv mbj_bandgap hse_gap  reference  \
0              na               na          na      na  mp-772283   
1          140.12            68.21          na      na    mp-2645   

        search                                               desc  
0  -Cr-Li-O-Sb  [2303.6175, 1627.410025, 5.445, -35.224125, 0....  
1       -Ni-Yb  [3435.925, 2327.6257, 9.72, -100.1516999999999...  

[2 rows x 62 columns]

keep only the necessary columns(id, structure data, target label)  
and divide 8:1:1 for training, validation and test.

In [31]:
target = "formation_energy_peratom"
ds = ds[["jid", "atoms", target]]

n_train = int(0.8*len(ds))
n_valid = int(0.1*len(ds))
train = ds[:n_train]
valid = ds[n_train:n_train+n_valid]
test = ds[n_valid+n_train:]

save them

In [32]:
train.to_pickle("./data/train_dft3d.pkl")
valid.to_pickle("./data/valid_dft3d.pkl")
test.to_pickle("./data/test_dft3d.pkl")

### crystal structure text (atoms, structure) -> graph, line graph  
It takes long time (pandas is in...) I recommend to do it separately (on local computer).  
In that case, after generate .data file which is including graphs,   
put them in ./data and tell the name(path) without the tail part "train(/valid/test/).data" in data_dir of ALIGNNDataModule. 

In [9]:
jarvis_data = ALIGNNDataModule(
    data_dir="./data/jarvis_dgl_",
    graph_fmt="dgl"
)

In [ ]:
jarvis_data.prepare_data(
    df_train=train,
    df_valid=valid,
    df_test=test,
    use_pandarallel=True
)

setup the data for training

In [10]:
jarvis_data.setup(
    stage="fit", 
    minmax_scale=False,
    standard_scale=True
    )

In [16]:
trainloader = jarvis_data.train_dataloader()
valloader = jarvis_data.val_dataloader()

### Training with pytorch_lightning

In [17]:
trainer = pl.Trainer(max_epochs=100,
                    # precision=16,
 
                    #profiler="pytorch",
                    callbacks=[
                        EarlyStopping(monitor="val_loss", mode="min", verbose=True, patience=10),
                        # StochasticWeightAveraging(swa_lrs=1e-2)
                        ]
                    )

INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:IPU available: False, using: 0 IPUs
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs


### fitting

In [ ]:
trainer.fit(
    model=ALIGNN(),
    train_dataloaders=trainloader,
    val_dataloaders=valloader

)

INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name            | Type       | Params
-----------------------------------------------
0 | atom_embedding  | MLPLayer   | 24.3 K
1 | edge_embedding  | Sequential | 22.5 K
2 | angle_embedding | Sequential | 19.9 K
3 | alignn_layers   | ModuleList | 1.3 M 
4 | gcn_layers      | ModuleList | 659 K 
5 | readout         | AvgPooling | 0     
6 | fc              | Linear     | 257   
-----------------------------------------------
2.0 M     Trainable params
0         Non-trainable params
2.0 M     Total params
8.187     Total estimated model params size (MB)


Sanity Checking: 0it [00:00, ?it/s]

Training: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved. New best score: 0.165


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.023 >= min_delta = 0.0. New best score: 0.142


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.021 >= min_delta = 0.0. New best score: 0.121


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.018 >= min_delta = 0.0. New best score: 0.103


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.009 >= min_delta = 0.0. New best score: 0.094


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.011 >= min_delta = 0.0. New best score: 0.083


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.010 >= min_delta = 0.0. New best score: 0.073


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.005 >= min_delta = 0.0. New best score: 0.067


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.002 >= min_delta = 0.0. New best score: 0.065


Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.006 >= min_delta = 0.0. New best score: 0.059


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.002 >= min_delta = 0.0. New best score: 0.056


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.008 >= min_delta = 0.0. New best score: 0.049


Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.003 >= min_delta = 0.0. New best score: 0.046


Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.005 >= min_delta = 0.0. New best score: 0.041


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.002 >= min_delta = 0.0. New best score: 0.039


Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.001 >= min_delta = 0.0. New best score: 0.039


Validation: 0it [00:00, ?it/s]

INFO:pytorch_lightning.callbacks.early_stopping:Metric val_loss improved by 0.003 >= min_delta = 0.0. New best score: 0.035


Validation: 0it [00:00, ?it/s]

In [ ]:
%reload_ext tensorboard
%tensorboard --logdir=lightning_logs/

save trainig results

In [ ]:
trainer.save_checkpoint("./trainer.chkpt")

load trainer

In [11]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
check_point = torch.load("/content/alignn_lightning/trainer.chkpt" ,map_location=torch.device(device))

In [12]:
model_weights = check_point["state_dict"]

In [13]:
model = ALIGNN()
model.load_state_dict(model_weights)

<All keys matched successfully>

In [14]:
jarvis_data.setup(
    stage="test",
    )
test_dataloader = jarvis_data.test_dataloader()

inference for test data

In [16]:
trainer = pl.Trainer()
predictions = trainer.predict(model, test_dataloader)

INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:IPU available: False, using: 0 IPUs
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Predicting: 0it [00:00, ?it/s]

The scaler coefficient is made at jarvis_data.setup at "fit" stage.  
If you started with load checkpoint, you need to execute setup with "fit".  

In [17]:
label_mean = torch.load("./data/jarvis_dgl_label_mean.data")
label_std = torch.load("./data/jarvis_dgl_label_std.data")

corrected_predictions = [pred * label_std + label_mean for pred in predictions]

In [18]:
mean_squared_error(jarvis_data.dataset_test.labels.numpy(), corrected_predictions, squared=False)

0.34101054

ok, MSE for formation energy inference is 0.34.  
I need to adjust hyperparameters for training and cross validation.